src/vfs2.c

```diff
@@ -171,75 +192,97 @@ struct vfs2_wal_index_full_hdr {
+	 * write to the WAL, which should be the header */
+	uint32_t page_size;
+
+	/* For ACTIVE, HIDDEN, POLLED: the header that hides the pending txn */
+	struct wal_index_basic_hdr prev_txn_hdr;
+	/* For ACTIVE, HIDDEN, POLLED: the header that shows the pending txn */
+	struct wal_index_basic_hdr pending_txn_hdr;
+
+	/* shm implementation; holds the WAL index */
+	void **shm_regions;
+	int shm_regions_len;
+	unsigned shm_refcount;
+	/* Zero for unlocked, positive for read-locked, UINT_MAX for write-locked */
+	unsigned shm_locks[SQLITE_SHM_NLOCK];
+
+	/* For ACTIVE, HIDDEN: the pending txn */
+	struct vfs2_wal_frame *pending_txn_frames;
+	uint32_t pending_txn_start;
 	uint32_t pending_txn_len;
-	uint32_t pending_txn_last_frame_commit; /* commit marker for the
-						   physical last frame */
-};
+	uint32_t pending_txn_last_frame_commit;
 
-struct vfs2_db {
-	const char *name; /* e.g. /path/to/my.db */
-
-	/* Copy of the WAL index header that reflects the last really-committed
-	 * (i.e. in Raft too) transaction, or the initial state of the WAL if
-	 * no transactions have been committed yet. */
-	struct vfs2_wal_index_basic_hdr prev_txn_hdr;
-	/* Copy of the WAL index header that reflects a sorta-committed
-	 * transaction that has not yet been through Raft, or all zeros
-	 * if no transaction fits this description. */
-	struct vfs2_wal_index_basic_hdr pending_txn_hdr;
-	/* When the WAL is restarted (or started for the first time), we capture
-	 * the initial WAL index header in prev_txn_hdr.
-	 *
-	 * When we get SQLITE_FCNTL_COMMIT_PHASETWO, we copy the WAL index
-	 * header from shm into pending_txn_hdr, then overwrite the shm with
-	 * prev_txn_hdr to hide the transaction.
-	 *
-	 * When we get vfs2_apply, we overwrite both prev_txn_hdr and the shm
-	 * with pending_txn_hdr. */
-
-	void **regions;
-	int regions_len;
-	unsigned refcount;
-
-	unsigned locks[SQLITE_SHM_NLOCK];
+	/* Frame index, points to the physical end of WAL-cur */
+	uint32_t wal_cursor;
+	/* Cached header of WAL-cur */
+	struct wal_hdr wal_cur_hdr;
+	/* Cached header of WAL-prev */
+	struct wal_hdr wal_prev_hdr;
+
+	struct sm wtx_sm;
+	/* VFS-wide data (immutable) */
+	struct common *common;
 };
 
 /**
  * VFS-specific file object, upcastable to sqlite3_file.
  */
-struct vfs2_file {
-	struct sqlite3_file base; /* vtable, must be first */
-	sqlite3_file *orig;       /* underlying file object */
-	struct vfs2_data *vfs_data;
-	struct vfs2_db_entry *entry;
-	int flags; /* from xOpen */
-	union {
-		/* if this file object is a WAL */
-		struct vfs2_wal wal;
-		/* if this file object is a main file */
-		struct vfs2_db db_shm;
-	};
+struct file {
+	/* Our custom sqlite3_io_methods vtable; must come first. Always
+	 * present. */
+	struct sqlite3_file base;
+	/* Flags passed to the xOpen that created this file. */
+	int flags;
+	/* File object created by the base (unix) VFS. Not used for WAL. */
+	sqlite3_file *orig;
```

Comment:
Needs to be a pointer since the size of `sqlite3_file` doesn't reflect the size of the file object created by the unix VFS -- it adds its own fields afterward just like we do with our `struct file`